### PR TITLE
Fix "TypeError: require(...).then is not a function" from shared bundle

### PR DESCRIPTION
**Pull request by wbinnssmith**
Fixes T-1100

Bug: Module exports are returned from the require transformed from import(...) in the case that the asset with the import has been extracted into a shared bundle, but is available synchronously up the graph or in the same bundle. 

This bug only occurs in production builds without scope hoisting.

This results in:

```
Uncaught (in promise) TypeError: o(...).then is not a function
    at Object.<anonymous> (a.bec2fa29.js:1)
    at o (index.js:1)
    at Object.<anonymous> (a.bec2fa29.js:1)
    at o (index.js:1)
    at Object.<anonymous> (:5000/dist/a.f08abe55.js:1)
    at o (index.js:1)
    at index.js:1
```

(where `o` is the `require` function Parcel passes to the module closure)

For now, I've changed optimize to also run its own internalization process, and this seems to have no meaningful impact in a large real-world app (though it may in cases with many async dependencies as this introduces more upward traversals). This should be addressed in the upcoming redesign of the bundler.

To do:
* [x] Fix the bug

Test Plan:
* [x] Added integration test

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.44s | +139.00ms |
| Cached | 507.00ms | +19.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.1e014c76.png | 274.00b | +0.00b | 251.00ms | -4.74s 🚀 |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.34s | +681.00ms ⚠️ |
| Cached | 461.00ms | -39.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 579.69kb | +0.00b | 6.17s | +539.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 579.69kb | +0.00b | 5.73s | -315.00ms 🚀 |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/304748109296042562)**

**Review by gorakong — APPROVED** [review]

File: packages/core/integration-tests/test/integration/shared-bundle-internalization/large.mjs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import "./async.mjs";
+import mod from "./imports-async.mjs";
+
+let text = "lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor lorem ipsum dolor";
```
Comment (mischnic) [inline packages/core/integration-tests/test/integration/shared-bundle-internalization/large.mjs]: One day, swc might decide to optimize this variable away, then it wouldn't be a shared bundle anymore.

Comment (wbinnssmith) [reply packages/core/integration-tests/test/integration/shared-bundle-internalization/large.mjs]: Yeah, I had the same concern and had a console.log here in an earlier iteration. Maybe we should have a global function that's a noop or something in the test env?

**Review by AGawrys — APPROVED** [review]

File: packages/bundlers/default/src/DefaultBundler.js [hunk 2]
```diff
@@ -451,6 +400,7 @@ export default (new Bundler({
 
     // Remove assets that are duplicated between shared bundles.
     deduplicate(bundleGraph);
+    internalizeReachableAsyncDependencies(bundleGraph);
```
Comment (devongovett) [inline packages/bundlers/default/src/DefaultBundler.js:403]: Any perf numbers to report on this extra step?

**Comment by wbinnssmith** [comment]
See the PR summary. In our app there was no significant difference (some samples ran faster with this actually)

**Review by devongovett — APPROVED** [review]

File: packages/bundlers/default/src/DefaultBundler.js (hunk 2 above)
Comment (wbinnssmith) [reply packages/bundlers/default/src/DefaultBundler.js:403]: (see below)

**Pull request merged**